### PR TITLE
feat: in app prompts experiment

### PR DESCRIPTION
**Pull request by kappa90**
## Problem

Following [this RFC](https://github.com/PostHog/product-internal/pull/318), this PR is an experimental implementation of customizable in-app prompts

## Changes

- There is a new endpoint, `PATCH /prompts/my_prompts`, which is used to receive the prompt sequences that should be shown to a person, as well as update and receive the state for each sequence (i.e. which prompts have already been displayed for the current person)
- A new logic `inAppPromptLogic` takes care of syncing the local person prompt state with the backend, as well as deciding which prompts to display
- For now, prompts are displayed conditionally based on the current path, using wildcards (e.g. `/dashboard/*`). We will run this as an experiment so we can use the experiment cohorts logic to show this conditionally to certain persons, but in the future this could be a native feature of prompts (by extending feature flags to carry a prompt configuration payload).
- Currently the only type of implemented prompt is a new type of tooltip called LemonActionableTooltip, which can be displayed next to UI elements, can have custom icons and actionable buttons (see screenshots below)
- The current flow is based on [this copy](https://docs.google.com/spreadsheets/d/1xLJeQhuQorjwDt4_L2vv71uq6wr0rSCIa0bCnGQ82DI/edit?usp=sharing)...
- ...however I feel like we need to make each sequence of tooltips independent of the navigation order a certain user might follow. We can't enforce users to discover the app in a specific way, so we should just create standalone product tours for each page, regardless of the direction they take
- There is a new button within the Help Menu, where you can re-trigger a product tour for any page where there is one.
- For the experiment, the plan is to run this only for users which sign-up from the start of the experimentation onwards

To-do:

- [ ] Finish writing kea tests
- [ ] @joethreepwood let's make the copy path independent, as mentioned above
- [ ] @clarkus we reviewed this together but take a look anyways

[Here's a video preview of the flow](https://www.loom.com/share/f2509b2bc6d54ddebe0254e58769d5ca)

Screenshot 2022-07-25 at 15 40 30
Screenshot 2022-07-25 at 15 40 59
Screenshot 2022-07-25 at 15 40 59
Screenshot 2022-07-25 at 15 40 46

## How did you test this code?

New tests for the frontend logic

### CODE REVIEW

**Comment by joesaunderson** [comment]
Heavily heavily heavily invested in this feature (from Mention Me, an enterprise customer).

File: frontend/src/lib/logic/inAppPrompt/inAppPromptLogic.tsx [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+                    actions.dismissSequence()
+                },
+                previous: actions.previousPrompt,
+                next: actions.nextPrompt,
+            })
+            cache.runOnClose = close
+
+            show.then(() => {
+                actions.updatePromptState({ step: values.currentStep })
+            }).catch((err) => console.error(err))
```
Comment (benjackwhite) [inline frontend/src/lib/logic/inAppPrompt/inAppPromptLogic.tsx:309]: Not super strongly opinionated on this but if we are in an async function perhaps we should keep things async-style and try-catch this?

File: frontend/src/lib/logic/inAppPrompt/inAppPromptLogic.tsx [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+                    break
+                default:
+                    break
+            }
+        },
+        //@ts-expect-error
+        [router.actions.locationChanged]: () => {
+            actions.closePrompts()
+            actions.findValidSequences()
+        },
```
Comment (benjackwhite) [inline frontend/src/lib/logic/inAppPrompt/inAppPromptLogic.tsx:438]: There's also `urlToAction` from kea-router which could listen to `*` route and trigger changes that way?

File: frontend/src/lib/logic/inAppPrompt/inAppPromptLogic.tsx [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+                    actions.syncState({ forceRun: true })
+                }
+            })
+        },
+        beforeUnmount: [
+            () => {
+                cache.runOnClose?.()
+            },
+        ],
+    })),
```
Comment (benjackwhite) [inline frontend/src/lib/logic/inAppPrompt/inAppPromptLogic.tsx:453]: I believe these have their own top-level builders so you could write it:

```suggestion
    afterMount(({ actions }) => {
        posthog.onFeatureFlags(async (_, variants) => {
            if (variants[FEATURE_FLAGS.IN_APP_PROMPTS_EXPERIMENT] === 'test') {
                actions.syncState({ forceRun: true })
            }
        })
    }),
    beforeUnmount(({ cache }) => cache.runOnClose?.()),
```

File: frontend/src/scenes/dashboard/DashboardHeader.tsx [hunk 4]
```diff
@@ -239,6 +239,7 @@ export function DashboardHeader(): JSX.Element | null {
                                     <LemonButton
                                         type="secondary"
                                         data-attr="dashboard-share-button"
+                                        data-tooltip="experiment-new-dashboard-product-tour-1"
```
Comment (benjackwhite) [inline frontend/src/scenes/dashboard/DashboardHeader.tsx:242]: Thinking to the future... We use `data-attr` everywhere and encourage people to do so for `autocapture`. Might be good to have that as a selectable option (thinking more about productizing the feature)

File: frontend/src/scenes/dashboard/Dashboards.tsx [hunk 5]
```diff
@@ -187,7 +189,15 @@ export function Dashboards(): JSX.Element {
             <PageHeader
                 title="Dashboards"
                 buttons={
-                    <LemonButton data-attr={'new-dashboard'} onClick={showNewDashboardModal} type="primary">
+                    <LemonButton
+                        data-attr={'new-dashboard'}
+                        data-tooltip="experiment-dashboards-product-tour"
+                        onClick={() => {
+                            closePrompts()
+                            showNewDashboardModal()
```
Comment (benjackwhite) [inline frontend/src/scenes/dashboard/Dashboards.tsx:197]: This is gonna be fun... We'll have to see how it goes with the trial but perhaps it shouldn't be possible to do anything other than interact with the tooltips as there are lots of other places that create modals...

File: posthog/api/prompt.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            my_prompts["sequences"].append(serialized_sequence)
+            if state:
+                my_prompts["state"][serialized_sequence["key"]] = state
+
+        if states_to_create:
+            PromptSequenceState.objects.bulk_create(states_to_create)
+        if states_to_update:
+            PromptSequenceState.objects.bulk_update(
+                states_to_update, ["last_updated_at", "step", "completed", "dismissed"]
+            )
```
Comment (benjackwhite) [inline posthog/api/prompt.py:100]: Totally fine for now but if we want to productionize this it's probably worth having this wrapped in a transaction, otherwise we could get into a weird state where half of your request worked. But probably overall would have a different implementation anyways so all good :D

**Review by benjackwhite — COMMENTED** [review]
Bunch of small/medium comments. Only thing really missing for approval would be some tests, especially around the api endpoint and the logic

**Comment by joethreepwood** [comment]
> * ...however I feel like we need to make each sequence of tooltips independent of the navigation order a certain user might follow. We can't enforce users to discover the app in a specific way, so we should just create standalone product tours for each page, regardless of the direction they take

Happy to draft up some copy for this and tooltip text v2 is on my current sprint...

However, I do think it could be worth revisiting the idea of a multi-page tutorial flow in the future, even if we just test it down the line. The main reason for this is that it'd give a chance to provide better context of how some tools can be used together contextually -- such as insights and dashboards, persons and session recordings, etc. I also think that, if we're confident what a majority of people want to achieve and can bake that into the flow, it could lead to the best result. 

Not a strong feeling for now, just something I'd like to consider for the future.

**Comment by joethreepwood** [comment]
> * The current flow is based on [this copy](https://docs.google.com/spreadsheets/d/1xLJeQhuQorjwDt4_L2vv71uq6wr0rSCIa0bCnGQ82DI/edit?usp=sharing)...
> * ...however I feel like we need to make each sequence of tooltips independent of the navigation order a certain user might follow. We can't enforce users to discover the app in a specific way, so we should just create standalone product tours for each page, regardless of the direction they take

Also, to confirm: If the tutorial flow is going to be independant then will skipping the tutorial skip the ENTIRE tutorial, or just the content for a single page?

**Comment by kappa90** [comment]
> > * ...however I feel like we need to make each sequence of tooltips independent of the navigation order a certain user might follow. We can't enforce users to discover the app in a specific way, so we should just create standalone product tours for each page, regardless of the direction they take
> 
> Happy to draft up some copy for this and tooltip text v2 is on my current sprint...
> 
> However, I do think it could be worth revisiting the idea of a multi-page tutorial flow in the future, even if we just test it down the line. The main reason for this is that it'd give a chance to provide better context of how some tools can be used together contextually -- such as insights and dashboards, persons and session recordings, etc. I also think that, if we're confident what a majority of people want to achieve and can bake that into the flow, it could lead to the best result.
> 
> Not a strong feeling for now, just something I'd like to consider for the future.

I think we just need to slightly tweak the copy so that one doesn't feel like they jumped in the middle of something.

From my experience, unless you force a flow, it's nearly impossible to dictate what the user does.

*Maybe* a solution to this is to have a check-list of product-tours the user can do, and have them complete them all. That way they can do them in whatever order, and we keep the copy focused on the current page.

> Also, to confirm: If the tutorial flow is going to be independant then will skipping the tutorial skip the ENTIRE tutorial, or just the content for a single page?

This would disable product tours in general, which can be then re-triggered from the Help Menu.

**Comment by joethreepwood** [comment]
For now, I've written up a draft of copy that gives 1-3 tooltip prompts for every main page in the app. It's a bit rough, with a focus on highlighting main functionality and features that may be missed (such as live events toggle). 

[Here it is.](https://docs.google.com/document/d/1lEylNEfe7FluuPsl1OhwL4F0bsQGEZMzbxchwhesoL4/edit?usp=sharing)

File: frontend/src/lib/logic/inAppPrompt/inAppPromptLogic.tsx [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+    step: number
+    type: PromptType
+    text: string
+    placement: Placement
+    reference: string
+    buttons: PromptButton[]
+    icon?: string
+}
+
+export type Tooltip = Prompt & { type: 'tooltip' }
```
Comment (pauldambra) [inline frontend/src/lib/logic/inAppPrompt/inAppPromptLogic.tsx:39]: #til

File: latest_migrations.manifest [hunk 8]
```diff
@@ -3,7 +3,7 @@ auth: 0012_alter_user_first_name_max_length
 axes: 0006_remove_accesslog_trusted
 contenttypes: 0002_remove_content_type_name
 ee: 0013_silence_deprecated_tags_warnings
-posthog: 0251_event_buffer
+posthog: 0252_auto_20220721_0858
```
Comment (pauldambra) [inline latest_migrations.manifest:6]: this will clash as well. I snuck in 252 ahead of you :)

Comment (pauldambra) [reply latest_migrations.manifest:6]: And we should be naming the migrations when creating them

File: posthog/api/prompt.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    authentication_classes = [
+        PersonalAPIKeyAuthentication,
+        TemporaryTokenAuthentication,  # Allows endpoint to be called from the Toolbar
+        authentication.SessionAuthentication,
+        authentication.BasicAuthentication,
+    ]
+
+    @action(methods=["PATCH"], detail=False)
+    def my_prompts(self, request: request.Request, **kwargs):
+        if not request.user.is_authenticated:  # for mypy
```
Comment (pauldambra) [inline posthog/api/prompt.py:45]: Can use a test to prove it but I'd expect this to be handled by middleware not each route

File: posthog/models/prompt.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+    last_updated_at: models.DateTimeField = models.DateTimeField(default=timezone.now)
+    step: models.IntegerField = models.IntegerField(default=1)
+    completed: models.BooleanField = models.BooleanField(default=False)
+    dismissed: models.BooleanField = models.BooleanField(default=False)
+
+
+# this model is currently unused as we are running an experimentation config, might be used in the future
+# or completely removed if we allow feature flags to carry a payload, which would be a prompts configuration JSON
+class PromptSequence(models.Model):
```
Comment (pauldambra) [inline posthog/models/prompt.py:25]: We can't delete models/fields for various reasons. If we're not using this we should remove it and only add it when we need it. Otherwise there's a risk it lives on in the code unused for a long time

**Review by pauldambra — COMMENTED** [review]
First pass over, looks good. Some nice use of TS types 💖 

Needs API tests and updating from master (but you know that :))

**Comment by kappa90** [comment]
@benjackwhite @pauldambra introduced all your feedback, we can merge Monday morning :)

File: frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.scss [hunk 11]
```diff
@@ -0,0 +1,43 @@
+.LemonActionableTooltip {
+    max-width: var(--in-app-prompts-width);
+    .LemonActionableTooltip__header {
+        display: flex;
```
Comment (pauldambra) [inline frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.scss:4]: **not blocking** but makes me wonder about our custom CSS metric... should we be extending utility classes in files like this 🤔

File: frontend/src/scenes/project-homepage/ProjectHomepage.tsx [hunk 12]
```diff
@@ -43,7 +44,7 @@ export function ProjectHomepage(): JSX.Element {
                 Invite members
             </LemonButton>
             <LemonButton
-                data-attr="project-home-new-insight"
```
Comment (benjackwhite) [inline frontend/src/scenes/project-homepage/ProjectHomepage.tsx:46]: I wonder if this is used anywhere in autocapture analytics....

File: frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.tsx [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    visible: boolean
+    close: () => void
+    element?: HTMLElement
+    next?: () => void
+    previous?: () => void
+    buttons?: { label: string; url?: string; action?: () => void }[]
+    icon?: JSX.Element
+}
+
+export const LemonActionableTooltip = ({
```
Comment (pauldambra) [inline frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.tsx:23]: should we insist on storybook stories for all components?

**not blocking** since it could be added in a follow-up

File: frontend/src/scenes/session-recordings/player/PlayerEvents.tsx [hunk 14]
```diff
@@ -118,6 +118,7 @@ export function PlayerEvents(): JSX.Element {
                     onClick={() => {
                         handleEventClick(event.playerPosition)
                     }}
+                    data-tooltip={'recording-event-list'}
```
Comment (benjackwhite) [inline frontend/src/scenes/session-recordings/player/PlayerEvents.tsx:121]: nit: doesn't need to be bracketed

**Review by benjackwhite — APPROVED** [review]
No big complaints only a few comments 👍

File: frontend/src/lib/logic/inAppPrompt/inAppPromptLogic.test.ts [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+            key: 'experiment-events-product-tour',
+            step: 0,
+            completed: false,
+            dismissed: false,
+            last_updated_at: '2022-07-26T16:32:55.153Z',
+        },
+    },
+}
+
+describe('inAppPromptLogic', () => {
```
Comment (pauldambra) [inline frontend/src/lib/logic/inAppPrompt/inAppPromptLogic.test.ts:73]: brilliant  to have such clear test names in here 🙌

File: posthog/api/prompt.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+class PromptSequenceStateViewSet(StructuredViewSetMixin, viewsets.ViewSet):
+    """
+    Create, read, update and delete prompt sequences state for a person.
+    """
+
+    serializer_class = PromptSequenceStateSerializer
+
+    @action(methods=["PATCH"], detail=False)
+    def my_prompts(self, request: request.Request, **kwargs):
+        if not request.user.is_authenticated:  # for mypy otherwise unauthenticated users don't have distinct_id
```
Comment (pauldambra) [inline posthog/api/prompt.py:28]: nit... should this be authenticated users in the comment (as opposed to unauthenticated)

File: frontend/src/scenes/project-homepage/ProjectHomepage.tsx (hunk 12 above)
Comment (kappa90) [reply frontend/src/scenes/project-homepage/ProjectHomepage.tsx:46]: This is definitely a mistake on my side, thanks for spotting that!

File: posthog/api/prompt.py [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+            )
+            local_states.append(parsed_state)
+
+        my_prompts: Dict[str, Any] = {"sequences": [], "state": {}}
+        states_to_update: List[PromptSequenceState] = []
+        states_to_create: List[PromptSequenceState] = []
+
+        person_id = (
+            PersonDistinctId.objects.filter(distinct_id=request.user.distinct_id, team_id=self.team_id)
+            .values_list("person_id", flat=True)
```
Comment (pauldambra) [inline posthog/api/prompt.py:43]: nice! select only what you need 💖

File: posthog/api/test/test_prompt.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+from freezegun.api import freeze_time
+from rest_framework import status
+
+from posthog.models import User
+from posthog.models.person.person import Person
+from posthog.models.prompt import PromptSequenceState, experiment_config
+from posthog.test.base import APIBaseTest
+
+
+class TestPrompt(APIBaseTest):
```
Comment (pauldambra) [inline posthog/api/test/test_prompt.py:12]: we should have a cross-team and cross user test similar to these https://github.com/PostHog/posthog/blob/master/posthog/api/test/test_insight.py#L1243

i.e. 

* user A can't view/amend prompts for user B
* user in Team A can't view/amend prompts for user in Team B

**Review by pauldambra — DISMISSED** [review]

File: frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.scss (hunk 11 above)
Comment (kappa90) [reply frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.scss:4]: and that's why something like tailwind was useful :D

I think one future goal should be to make custom classes just a combination of utility classes

File: frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.tsx (hunk 13 above)
Comment (kappa90) [reply frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.tsx:23]: Good point, I will do the stories for these in a separate PR

File: posthog/api/prompt.py (hunk 16 above)
Comment (kappa90) [reply posthog/api/prompt.py:28]: The double negation makes it confusing I guess, let me rephrase it

If we don't do a type check on authenticated users, mypy will throw an error because unauthenticated ones miss the distinct_id, which is required just after.

File: posthog/api/test/test_prompt.py (hunk 18 above)
Comment (kappa90) [reply posthog/api/test/test_prompt.py:12]: Brilliant, caught a bug thanks to this, added a test, thanks

File: frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.tsx [hunk 19]
```diff
@@ -0,0 +1,110 @@
+import React from 'react'
+import { Placement } from '@floating-ui/react-dom-interactions'
+import { Popup } from 'lib/components/Popup/Popup'
+import { IconOpenInNew } from 'lib/components/icons'
+import { CloseOutlined, LeftOutlined, RightOutlined } from '@ant-design/icons'
```
Comment (Twixes) [inline frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.tsx:5]: Can we use `IconClose`, `IconChevronLeft`, and `IconChevronRight` instead?

File: frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.tsx [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+                            {maxSteps > 1 && (
+                                <>
+                                    <LemonButton
+                                        className="LemonActionableTooltip__navigation--left"
+                                        onClick={previous}
+                                        disabled={step === 0}
+                                        size="small"
+                                        type="stealth"
+                                    >
+                                        <LeftOutlined />
```
Comment (Twixes) [inline frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.tsx:61]: Ideally icons should be put into `icon={}`

**Review by Twixes — DISMISSED** [review]
AntD alert

File: posthog/models/prompt.py [hunk 21]
```diff
@@ -0,0 +1,520 @@
...
+
+
+class PromptSequenceState(models.Model):
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(fields=["team", "person", "key"], name="unique sequence key for person for team")
+        ]
+
+    team: models.ForeignKey = models.ForeignKey("Team", on_delete=models.CASCADE)
+    person: models.ForeignKey = models.ForeignKey("Person", on_delete=models.CASCADE)
```
Comment (Twixes) [inline posthog/models/prompt.py:14]: As a user, I don't think it'd make sense for me to go through a flow multiple times just because I have multiple projects. Why not key this just `["person", "key"]`?

File: posthog/api/prompt.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+            .first()
+        )
+
+        if not person_id:
+            return Response("", status=status.HTTP_404_NOT_FOUND)
```
Comment (Twixes) [inline posthog/api/prompt.py:48]: This won't have a JSON payload, I believe we should raise a `exceptions.NotFound()` (not sure actually, since this is an `@action`, but we definitely should return the standard 404 JSON response)

File: posthog/api/prompt.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+        saved_states = PromptSequenceState.objects.filter(team=self.team, person_id=person_id)
+        all_sequences = get_active_prompt_sequences()
+
+        new_states: List[Dict] = []
+
+        for sequence in all_sequences:
+            local_state = next((s for s in local_states if sequence["key"] == s["key"]), None)
+            saved_state: Union[PromptSequenceState, None] = next(
```
Comment (Twixes) [inline posthog/api/prompt.py:57]: Tip: `Union[T, None]` can be represented as `Optional[T]` ;)

File: frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.tsx (hunk 19 above)
Comment (kappa90) [reply frontend/src/lib/components/LemonActionableTooltip/LemonActionableTooltip.tsx]: Screenshot 2022-08-01 at 10 29 15

Great thanks, removed them!

File: posthog/api/prompt.py [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+
+from posthog.api.routing import StructuredViewSetMixin
+from posthog.models.person.person import PersonDistinctId
+from posthog.models.prompt import PromptSequenceState, get_active_prompt_sequences
+
+
+class PromptSequenceStateSerializer(serializers.HyperlinkedModelSerializer):
+    class Meta:
+        model = PromptSequenceState
+        fields = ["key", "last_updated_at", "step", "completed", "dismissed"]
```
Comment (Twixes) [inline posthog/api/prompt.py:16]: Should they all also be marked as `read_only_fields`?

File: posthog/api/__init__.py [hunk 25]
```diff
@@ -57,6 +58,7 @@ def api_not_found(request):
 )
 projects_router.register(r"annotations", annotation.AnnotationsViewSet, "project_annotations", ["team_id"])
 projects_router.register(r"feature_flags", feature_flag.FeatureFlagViewSet, "project_feature_flags", ["team_id"])
+projects_router.register(r"prompts", prompt.PromptSequenceStateViewSet, "project_feature_flags", ["team_id"])
```
Comment (Twixes) [inline posthog/api/__init__.py:61]: `basename` needs to be updated
```suggestion
projects_router.register(r"prompts", prompt.PromptSequenceStateViewSet, "project_prompts", ["team_id"])
```

As for the shape of the API though, wouldn't it be more intuitive for prompts to be scoped per user instead per project? (mentioned this in a comment above too)

File: frontend/src/scenes/annotations/index.tsx [hunk 26]
```diff
@@ -100,14 +100,20 @@ export function Annotations(): JSX.Element {
                 title="Annotations"
                 caption="Here you can add organization- and project-wide annotations. Dashboard-specific ones can be added directly in the dashboard."
                 buttons={
-                    <LemonButton type="primary" data-attr="create-annotation" onClick={(): void => setOpen(true)}>
+                    <LemonButton
+                        type="primary"
+                        data-attr="create-annotation"
+                        data-tooltip="annotations-new-button"
```
Comment (Twixes) [inline frontend/src/scenes/annotations/index.tsx:106]: Could we just reuse `data-attr` instead of adding `data-tooltip`? 🤔 Seems like they serve a very similar purpose

File: frontend/src/lib/constants.tsx [hunk 27]
```diff
@@ -118,6 +118,7 @@ export const FEATURE_FLAGS = {
     // Re-enable person modal CSV downloads when frontend can support new entity properties
     PERSON_MODAL_EXPORTS: 'person-modal-exports', // hot potato see https://github.com/PostHog/posthog/pull/10824
     BILLING_LOCK_EVERYTHING: 'billing-lock-everything', // owner @timgl
+    IN_APP_PROMPTS_EXPERIMENT: 'IN_APP_PROMPTS_EXPERIMENT', // owner: @kappa90
```
Comment (Twixes) [inline frontend/src/lib/constants.tsx:121]: Absolute nit… but could we keep the same `kebab-case` convention for feature flag keys?

File: frontend/src/lib/components/icons.tsx [hunk 28]
```diff
@@ -1548,3 +1548,14 @@ export function IconLink(props: React.SVGProps<SVGSVGElement>): JSX.Element {
         </svg>
     )
 }
+
+export function IconMessages(props: React.SVGProps<SVGSVGElement>): JSX.Element {
+    return (
+        <svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg" {...props}>
```
Comment (Twixes) [inline frontend/src/lib/components/icons.tsx:1554]: This doesn't seem right – all our icons should be using a 24x24 viewbox.

File: frontend/src/lib/components/icons.tsx [hunk 29]
```diff
@@ -1548,3 +1548,14 @@ export function IconLink(props: React.SVGProps<SVGSVGElement>): JSX.Element {
...
+            <path
+                d="M13 2V9H3.17L2 10.17V2H13ZM14 0H1C0.45 0 0 0.45 0 1V15L4 11H14C14.55 11 15 10.55 15 10V1C15 0.45 14.55 0 14 0ZM19 4H17V13H4V15C4 15.55 4.45 16 5 16H16L20 20V5C20 4.45 19.55 4 19 4Z"
+                fill="#747EA2"
```
Comment (Twixes) [inline frontend/src/lib/components/icons.tsx:1557]: Icons needs to have `fill="currentColor"` so that they can adjust to button colors
